### PR TITLE
Feature: Add FastAP Loss 

### PR DESCRIPTION
**Pull request by KarahanS**
* Implementation of `FastAP` loss. Primarily inspired by [Pytorch-metric-learning](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/fast_ap_loss.py) implementation and [official implementation](https://github.com/kunhe/FastAP-metric-learning/blob/master/pytorch/FastAP_loss.py). For more details, refer to the paper https://cs-people.bu.edu/fcakir/papers/fastap_cvpr2019.pdf.
* Additionally, I updated the constructor of `Circle Loss` to give `distance_metric_name` to the constructor of the superclass. (I'm not sure if it was not given on purpose or a small mistake - [related PR](https://github.com/qdrant/quaterion/pull/197))
* Additionally, updated `get_anchor_negative_mask` and `get_anchor_positive_mask` to accept `labels_b` as `Optional[Tensor]` so that we can call those functions with only one parameter (as I did in FastAP Loss).
* I added official implementation to the `test_fast_ap_loss.py` so that we can compare out loss with the "expected" loss value (Inspired by the [test file](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/691a6354be130547a8e26e170b86cf65c36cd791/tests/losses/test_fastap_loss.py) written for pytorch-metric-learning.)

**This PR is now open for review. FYI @generall @monatis.**
I'd be happy to hear any suggestions, corrections, or improvements.

### CODE REVIEW

**Review by monatis — CHANGES_REQUESTED** [review]
Could you please add a test for this loss as well? You can also run an experiment by using it one of the examples in the repository.

There are two minor issues in the implementation that I commented, otherwise it looks promising.

File: quaterion/loss/fast_ap_loss.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        n_pos = torch.sum(pos_mask, dim=1)  # Sum over all columns (for each row)
+
+        # 2. compute distances from embeddings squared Euclidean distance matrix
+        embeddings = F.normalize(embeddings, p=2, dim=1)  # normalize embeddings
+        dist_matrix = (
+            self.distance_metric.distance_matrix(embeddings) ** 2
+        )  # (batch_size, batch_size)
+
+        # 3. estimate discrete histograms
+        # TODO: Learn if distance power can be greater than 2. If so, maximum histogram value can be greater than 4.
```
[the remark this answers is not in the document]
Comment (KarahanS) [reply quaterion/loss/fast_ap_loss.py:86]: Actually, I was going to ask that question to you. Right now, we hardcode the maximum histogram value as 4.0 (similar to the official implementation). I guess, there is no problem with that, right?

**Review by monatis — APPROVED** [review]
Thanks for handling these reviews. LGTM, I'll merge it.
